### ISSUE TITLE
Inferred return type intersects returned objects but only if they are not assigned in the function body

### ISSUE THREAD

**Issue by Aetherall** [microsoft/TypeScript#56692]
### 🔎 Search Terms

return type intersection
return type intersection undefined
return type object union
return type intersection narrowing

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, from 4.0.0 to 5.3

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.2#code/GYVwdgxgLglg9mABMAhjANgCgJQG8BQiRiMwmAjNogE4CmUI1SuiKAXIlNSLYgL6FidBk0QsARh1ToAzrwED8S1BnwB6NYgB6AfmXho8JAHc41ANYycBYiTKUbt4hAQyoNWjJDp3AXjGsHFw8-IJOHiJIdF4+YaG2LmBuHjF+AZKc3PJxwoxRnt5Q+IqmFjLqmrpAA

### 💻 Code

![image](https://github.com/microsoft/TypeScript/assets/13298494/43d43c20-ca37-4f53-a7bd-9e886cfbc5ec)

### 🙁 Actual behavior

```ts
function fail(){
    if(1) return { a: true }
    return { b: false }
}

fail
// ^? function fail(): { a: boolean; b?: undefined; } | { b: boolean; a?: undefined; }

function works(){
    if(1){
        const result = { a: true }
        return result
    }
    const result = { b: true }
    return result
}

works
// ^? function works(): { a: boolean; } | { b: boolean; }
```

### 🙂 Expected behavior

```ts
function fail(){
    if(1) return { a: true }
    return { b: false }
}

fail
// ^? function fail(): { a: boolean; } | { b: boolean; }

function works(){
    if(1){
        const result = { a: true }
        return result
    }
    const result = { b: true }
    return result
}

works
// ^? function works(): { a: boolean; } | { b: boolean; }
```

**Comment by jcalz** [comment]
Duplicate of #50516

**Comment by Aetherall** [comment]
The issue/PR referenced are supposedly solved/merged, but I as indicated, we can observe this behavior in latest typescript versions.
The PR https://github.com/microsoft/TypeScript/pull/19513 added a workaround for identifying properties that may or may not be accessible, but this issue highlights the incomplete implementation of the workaround. We either need to apply the normalisation all the way, or maybe to allow property access if one of the union member has such property then narrow down the value type on access. Either way, the fact we have two different behavior for virtually identical functions may be something to look into a bit further.

What do you think ?

**Comment by MartinJohns** [comment]
> The issue/PR referenced are supposedly solved/merged, but I as indicated, we can observe this behavior in latest typescript versions.

Because it's working as intended. That's why the issue I linked has the label "Working as Intended".

**Comment by Aetherall** [comment]
I took the time to read the referenced PRs this weekend, and I've noticed the labeling on the issue, but it seems more like its a side effect of a solution used to solve another use case :

The PR https://github.com/microsoft/TypeScript/pull/19513 addresses the ability to access properties. The intended behavior in that case is that the access is authorized and his types correctly narrowed down. However, a side effect to the solution used to address the https://github.com/microsoft/TypeScript/pull/19513 leads to this issue, which highlight an inconsistency when using a pattern that should not affect the typing.

For instance, this would also be a coherent behavior
```ts
function fail(){
    if(1) return { a: true }
    return { b: false }
}

fail
// ^? function fail(): { a: boolean; b?: undefined; } | { b: boolean; a?: undefined; }

function works(){
    if(1){
        const result = { a: true }
        return result
    }
    const result = { b: true }
    return result
}

works
// ^? function works(): { a: boolean; b?: undefined; } | { b: boolean; a?: undefined; }
```
even if less desirable, this solution falls along the same lines as the https://github.com/microsoft/TypeScript/pull/19513 PR solution.

The point in this issue is that differing types for the same functions really looks like a bug, or at least counter intuitive.

I am aware that it is probably not a big deal for most typescript developers, and I can use the variable assignation workaround for my use cases (it is no pretty code but will work fine). But it feels like fixing this with the lazy evaluation method really goes into the same direction as the recent improvements on type inference on conditions and switch statements.

Thanks for taking the time to read me :)
